pkg/services/ngalert/writer/datasourcewriter.go

```diff
@@ -78,6 +77,73 @@ func (w *DatasourceWriter) decrypt(ds *datasources.DataSource) (map[string]strin
 	return decryptedJsonData, err
 }
 
+func getPrometheusType(ds *datasources.DataSource) string {
+	if ds.JsonData == nil {
+		return ""
+	}
+	jsonData := ds.JsonData.Get("prometheusType")
+	if jsonData == nil {
+		return ""
+	}
+	str, err := jsonData.String()
+	if err != nil {
+		return ""
+	}
+	return str
+}
+
+func getRemoteWriteURL(ds *datasources.DataSource) (*url.URL, error) {
+	u, err := url.Parse(ds.URL)
+	if err != nil {
+		return nil, err
+	}
+
+	if getPrometheusType(ds) == "Prometheus" {
+		return u.JoinPath("/api/v1/write"), nil
+	}
+
+	// All other cases assume Mimir/Cortex, as these systems are much more likely to be
+	// used as a remote write target, where as Prometheus does not recommend it.
+
+	// Mimir/Cortex are more complicated, as Grafana has to be configured with the
+	// base URL for where the Prometheus API is located, e.g. /api/prom or /prometheus.
+	//
+	// - For "legacy" routes, /push is located on the same level as /api/v1/query.
+	//
+	//   For example:
+	//     Grafana will be configured with <host>/api/prom
+	//     The query API is at /api/prom/api/v1/query
+	//     The push API is at /api/prom/push
+	//
+	// - For "new" routes, /push is located at the Mimir root, not Prometheus root.
+	//
+	//   For example:
+	//     Grafana will be configured with e.g. <host>/prometheus
+	//     The query API is at /prometheus/api/v1/query
+	//     But push API is at /push
+	//
+	// Unfortunately, the prefixes can also be configured,
+
+	cleanPath := path.Clean(u.Path)
+
+	// If the suffix is /api/prom, assume Mimir/Cortex with legacy routes.
+	if strings.HasSuffix(cleanPath, "/api/prom") {
+		u.Path = path.Join(u.Path, "/push")
+		return u, nil
+	}
+
+	// If the suffix is /prometheus, assume Mimir/Cortex with new routes.
+	if strings.HasSuffix(cleanPath, "/prometheus") {
+		u.Path = path.Join(path.Dir(u.Path), "/api/v1/push")
+		return u, nil
+	}
+
+	// The user has configured an unknown prefix, so fall back to taking
+	// the host as the Mimir root. This is less than ideal.
```

Comment:
WDYT about instead not offering a data source as a target if we don't recognize its root until we have a better solution?

Reply:
This would cause problems for migration, but maybe the same answer applies there (i.e. don't migrate unless we know we'll have a valid recording rule target).

Reply:
It's a good idea, unfortunately we don't have any plumbing for telling the frontend what data sources are "usable" as it stands, so we'll have to live with it for now.